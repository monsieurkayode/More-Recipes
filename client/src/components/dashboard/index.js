/* jshint esversion: 6 */
import UserProfile from './UserProfile';
import UserRecipe from './UserRecipe';
import WelcomeDisplay from './WelcomeDisplay';
import DashboardPanel from './DashboardPanel';
import SideNavDashboard from './SideNavDashboard';
import UserFavoriteRecipe from './UserFavoriteRecipe';
import UserNotification from './UserNotification';

export {
  UserProfile,
  UserRecipe,
  UserFavoriteRecipe,
  UserNotification,
  WelcomeDisplay,
  SideNavDashboard,
  DashboardPanel
};
