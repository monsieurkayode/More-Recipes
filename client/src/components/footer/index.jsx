import React from 'react';

/**
 * Footer
 * @function Footer
 *
 * @param {void} void
 *
 * @return {JSX} JSX
 */
const Footer = () => (
  <footer className="page-footer blue-grey darken-2">
    <div className="row">
      <div className="col l3 social-icon-container">
        <ul>
          <li>
            <div>
              <span>
                <i className="fa fa-2x fa-github teal-text" />
              </span>
            </div>
          </li>
          <li>
            <div>
              <span>
                <i className="fa fa-2x fa-facebook teal-text" />
              </span>
            </div>
          </li>
          <li>
            <div>
              <span>
                <i className="fa fa-2x fa-twitter teal-text" />
              </span>
            </div>
          </li>
          <li>
            <div>
              <span>
                <i className="fa fa-2x fa-google teal-text" />
              </span>
            </div>
          </li>
        </ul>
      </div>
    </div>
    <div className="footer-copyright">
      <div className="container">
        Â© Andela 2018
        <a
          className="grey-text text-lighten-4 right"
          href="/"
        >
        Created by Kayode Adeola
        </a>
      </div>
    </div>
  </footer>
);

export default Footer;
