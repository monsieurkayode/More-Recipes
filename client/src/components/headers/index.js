/* jshint esversion: 6 */
import Banner from './Banner';
import DashboardNavbar from './DashboardNavbar';
import HomeNavbar from './HomeNavbar';
import LandingNavbar from './LandingNavbar';
import UserMenu from './UserMenu';
import SearchBar from './SearchBar';

export {
  Banner,
  DashboardNavbar,
  HomeNavbar,
  LandingNavbar,
  UserMenu,
  SearchBar
};
