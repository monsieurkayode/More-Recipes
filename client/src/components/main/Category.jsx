import React, { Component } from 'react';
import { connect } from 'react-redux';
import debounce from 'lodash/debounce';
import PropTypes from 'proptypes';
import { Link } from 'react-router-dom';

import { searchPost } from '../../actions';
import cleanString from '../../../../shared/cleanString';
import Categories from '../../../../shared/categories';
import pascalCase from '../../utils/pascalCase';

/**
 * @summary - Category class declaration
 * @class Category
 * @extends {Component}
 */
export class Category extends Component {
  /**
   * Component constructor
   * @param {object} props
   * @memberOf Category
   */
  constructor(props) {
    super(props);
    this.state = {
      searchType: '',
      searchTerm: ''
    };
  }

  /**
   * @method componentDidMount
   *
   * @param {void} void
   *
   * @returns {void}
   */
  componentDidMount() {
    $((this.search)).on('change', this.handleSearchType);
  }

  /**
   * @method componentWillUpdate
   *
   * @param {void} void
   *
   * @returns {void}
   */
  componentWillUpdate() {
    $('select').material_select();
  }

  /**
   * Handle search
   * @method handleSearch
   *
   * @param {void} void
   *
   * @returns {void}
   */
  handleSearch = () => {
    const { searchType, searchTerm } = this.state;
    if ((cleanString(searchTerm)).length % 3 === 0) {
      this.props.searchPost(searchType, searchTerm);
    }
  }

  /**
   * Handle change
   * @method handleChange
   *
   * @param {object} event
   *
   * @returns {void}
   */
  handleChange = (event) => {
    this.setState({
      [event.target.name]: event.target.value
    });
  }

  /**
   * Handle search criteria
   * @method handleSearchType
   *
   * @param {object} event
   *
   * @returns {void}
   */
  handleSearchType = (event) => {
    this.setState({
      searchType: event.target.value
    });
  }

  /**
   * @method renderCategory
   *
   * @param {number} index
   *
   * @returns {JSX} JSX
   */
  renderCategory = (index) => {
    const category = Categories[index];
    return (
      <li key={category}>
        <Link
          to={`/category/${category}`}
          className="collapsible-header"
        >
          {pascalCase(category)}
        </Link>
      </li>
    );
  }

  /**
   * Renders the component
   * @method render
   *
   * @returns {JSX} JSX
   */
  render() {
    const handleSearch = debounce(() => { this.handleSearch(); }, 300);
    return (
      <div className="col l2 offset-l1 m2 s12">
        <form id="search-form" onChange={handleSearch}>
          <input
            onChange={this.handleChange}
            value={this.state.searchTerm}
            className="white input-fa"
            name="searchTerm"
            type="text"
            placeholder="&#xf002; Enter search keyword..."
          />
          <select
            ref={(ref) => { this.search = ref; }}
            value={this.state.searchType}
          >
            <option value="" disabled>Search criteria</option>
            <option value="name">Title</option>
            <option value="category">Category</option>
            <option value="ingredients">Ingredients</option>
          </select>
        </form>
        <div className="category hide-on-small-only">
          <div>
            <div>
              <b>Featured Categories</b>
              <p className="divider" />
            </div>
            <ul className="collapsible z-depth-0">
              {Categories.slice(6).map(
                (category, index) => this.renderCategory(index)
              )}
            </ul>
          </div>
        </div>
      </div>
    );
  }
}

Category.propTypes = {
  searchPost: PropTypes.func.isRequired
};

export default connect(null, { searchPost })(Category);
