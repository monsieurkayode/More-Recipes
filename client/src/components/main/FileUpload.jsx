import React from 'react';
import PropTypes from 'proptypes';
/**
 * Hanldes file selection for upload
 *
 * @function handleChange
 *
 * @param {function} handler
 *
 * @returns {object} selected file
 */
const handleChange = handler => ({ target: { files } }) =>
  handler(files.length ? { file: files[0], name: files[0].name } : {});

const FileUpload = ({
  input: { onChange, onBlur, value: omitValue, ...inputProps },
  meta: omitMeta,
  ...props
}) => (
  <div id="upload-btn">
    <label htmlFor="image">
      <input
        type="file"
        id="image"
        onChange={handleChange(onChange)}
        onBlur={handleChange(onBlur)}
        placeholder="Upload Image"
        {...inputProps}
        {...props}
      />
      <span
        className="btn"
      >
        Upload Image
      </span>
    </label>
  </div>
);

FileUpload.propTypes = {
  input: PropTypes.shape({
    onChange: PropTypes.func.isRequired,
    onBlur: PropTypes.func.isRequired
  }).isRequired,
  meta: PropTypes.shape({}).isRequired
};

export default FileUpload;
