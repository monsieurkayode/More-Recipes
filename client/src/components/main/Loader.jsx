import React from 'react';
import PreloaderIcon, { ICON_TYPE } from 'react-preloader-icon';

/**
 * Loader
 * @function Loader
 *
 * @param {void} void
 *
 * @return {JSX} JSX
 */
const Loader = () => (
  <div className="loader">
    <PreloaderIcon
      type={ICON_TYPE.OVAL}
      size={80}
      strokeWidth={7}
      strokeColor="#006664"
      duration={800}
    />
  </div>
);

export default Loader;
