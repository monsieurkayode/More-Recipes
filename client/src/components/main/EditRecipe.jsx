import React, { Component } from 'react';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { Field, reduxForm } from 'redux-form';
import PropTypes from 'proptypes';

import FileUpload from './FileUpload';
import Loader from './Loader';
import { editPost, fetchSingleRecipe, isFetching } from '../../actions';
import validate from '../../utils/validateRecipe';
import categories from '../../../../shared/categories';
import pascalCase from '../../utils/pascalCase';

/**
 * @summary - EditRecipe class declaration
 * @class EditRecipe
 * @extends {Component}
 */
export class EditRecipe extends Component {
  /**
   * Component constructor
   * @param {object} props
   * @memberOf EditRecipe
   */
  constructor(props) {
    super(props);
    this.state = {
      selectedCategory: '',
    };
  }

  /**
   * @method componentDidMount
   *
   * @param {void} void
   *
   * @returns {void}
   */
  componentDidMount() {
    const { recipeId } = this.props.match.params;
    $(('#category')).on('change', this.handleCategory);
    this.props.fetchSingleRecipe(recipeId);
  }

  /**
   * @method componentWillReceiveProps
   *
   * @param {object} nextProps
   *
   * @returns {void}
   */
  componentWillReceiveProps(nextProps) {
    if (nextProps.initialValues.category) {
      this.setState({
        selectedCategory: nextProps.initialValues.category
      });
    }
  }

  /**
   * @method componentWillUpdate
   *
   * @param {void} void
   *
   * @returns {void}
   */
  componentWillUpdate() {
    $('select').material_select();
  }

  /**
   * Handle submit
   *
   * @method onSubmit
   *
   * @param {object} values
   * @param {string} category
   *
   * @returns {void}
   */
  onSubmit = (values, category) => {
    const { recipeId } = this.props.match.params;
    category = this.state.selectedCategory;
    this.props.isFetching(true, 'EditRecipe');
    this.props.editPost(recipeId, category, values, (message) => {
      this.props.history.goBack();
      setTimeout(() => Materialize.toast(message, 4000, 'grey darken-2'), 3000);
    });
  }

  /**
   * Handle category selection
   *
   * @method handleCategory
   *
   * @param {object} event
   *
   * @returns {void}
   */
  handleCategory = (event) => {
    this.setState({
      selectedCategory: event.target.value
    });
  }

  /**
   * @method renderInput
   *
   * @param {object} field
   *
   * @returns {JSX} JSX
   */
  renderInput = (field) => {
    const { label, input, meta, type, className, placeholder } = field;
    return (
      <div className="row">
        <div className="input-field col-s12 l12 m12">
          <input
            type={type}
            placeholder={placeholder}
            {...input}
            className={className}
          />
          <label htmlFor className="active">{label}</label>
        </div>
        <div className="red-text right">
          {meta.touched ? meta.error : ''}
        </div>
      </div>
    );
  }

  /**
   * @method renderTextArea
   *
   * @param {object} field
   *
   * @returns {JSX} JSX
   */
  renderTextArea = (field) => {
    const { label, input, meta, type, className, placeholder } = field;
    return (
      <div className="row">
        <div className="input-field col-s12 l12 m12">
          <textarea
            placeholder={placeholder}
            type={type}
            {...input}
            className={className}
          />
          <label htmlFor className="active">{label}</label>
        </div>
        <div className="red-text right">
          {meta.touched ? meta.error : ''}
        </div>
      </div>
    );
  }

  /**
   * @method renderCategory
   *
   * @param {object} field
   *
   * @returns {JSX} JSX
   */
  renderCategory = (field) => {
    const { value } = field;
    return (
      <div className="row">
        <div className="input-field col-s12 l12 m12">
          <select
            id="category"
            value={value}
          >
            <option value={this.state.selectedCategory}>Select Category</option>
            {categories
              .map(
                category => (<option value={category} key={category}>
                  {pascalCase(category)}
                </option>)
              )
            }
          </select>
        </div>
      </div>
    );
  }

  /**
   * Renders the component
   * @method render
   *
   * @returns {JSX} JSX
   */
  render() {
    const { handleSubmit, invalid, initialValues, isLoading } = this.props;
    return (
      <div className="container">
        <div className="row">
          {isLoading ?
            <Loader /> :
            <form
              className="col l8 m8 s10 offset-s1 offset-m2 offset-l2"
              onSubmit={handleSubmit(this.onSubmit)}
            >
              <Field
                type="text"
                className=""
                placeholder="Enter recipe name"
                label="Recipe Name"
                name="recipeName"
                component={this.renderInput}
              />

              <Field
                name="category"
                value={this.state.selectedCategory}
                component={this.renderCategory}
              />

              <Field
                type="text"
                placeholder="Markdown supported"
                className="materialize-textarea"
                label="Ingredients"
                name="ingredients"
                component={this.renderTextArea}
              />

              <Field
                type="text"
                placeholder="Markdown supported"
                className="materialize-textarea"
                label="Intsructions"
                name="instructions"
                component={this.renderTextArea}
              />
              <div>
                <Field
                  name="image"
                  component={FileUpload}
                />
                <span className="image-name">
                  <p>
                    {!!initialValues.image && initialValues.image.name }
                  </p>
                </span>
              </div>

              <div className="form-btn-control">
                <button
                  className="btn red"
                  onClick={() => this.props.history.push('/dashboard/recipes')}
                >
                Cancel
                </button>
                <button
                  disabled={invalid}
                  type="submit"
                  className="btn"
                >
                Update
                </button>
              </div>
            </form>}
        </div>
      </div>
    );
  }
}

EditRecipe.propTypes = {
  editPost: PropTypes.func.isRequired,
  fetchSingleRecipe: PropTypes.func.isRequired,
  initialValues: PropTypes.shape({
    id: PropTypes.number,
    views: PropTypes.number,
    upvote: PropTypes.number,
    downvote: PropTypes.number,
    recipeName: PropTypes.string,
    category: PropTypes.string,
    ingredients: PropTypes.string,
    instructions: PropTypes.string,
    image: PropTypes.shape({})
  }).isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
    goBack: PropTypes.func.isRequired,
  }).isRequired,
  handleSubmit: PropTypes.func.isRequired,
  invalid: PropTypes.bool.isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      recipeId: PropTypes.string.isRequired
    }).isRequired
  }).isRequired,
  isLoading: PropTypes.bool.isRequired,
  isFetching: PropTypes.func.isRequired
};

const mapStateToProps = ({ currentRecipe, isLoading }) => ({
  initialValues: {
    recipeName: currentRecipe.recipeName,
    category: currentRecipe.category,
    ingredients: currentRecipe.ingredients,
    instructions: currentRecipe.instructions,
    image: {
      file: 'empty',
      name: currentRecipe.image
    }
  },
  isLoading: isLoading.editRecipeIsLoading
});

export default compose(
  connect(mapStateToProps, { editPost, fetchSingleRecipe, isFetching }),
  reduxForm({
    validate,
    form: 'EditRecipeForm',
    enableReinitialize: true,
  }),
)(EditRecipe);
