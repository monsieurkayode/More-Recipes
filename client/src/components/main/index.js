/* jshint esversion: 6 */
import Category from './Category';
import Recipes from './Recipes';
import RecipeItem from './RecipeItem';
import TopRecipeItem from './TopRecipeItem';
import TopRecipes from './TopRecipes';
import SignupForm from './SignupForm';
import SigninForm from './SigninForm';
import TextField from './TextField';
import PostRecipe from './PostRecipe';
import EditRecipe from './EditRecipe';
import SideNav from './SideNav';
import WelcomeMessage from './WelcomeMessage';
import Loader from './Loader';

export {
  Category,
  Recipes,
  RecipeItem,
  TopRecipeItem,
  TopRecipes,
  PostRecipe,
  EditRecipe,
  SignupForm,
  SigninForm,
  TextField,
  SideNav,
  WelcomeMessage,
  Loader
};
