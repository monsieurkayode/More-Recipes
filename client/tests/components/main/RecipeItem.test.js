/* eslint-disable no-undef */
import React from 'react';

import { RecipeItem } from '../../../src/components/main';
import data from '../../__mocks__/mockData';

const setup = () => {
  const props = {
    grid: 'm4',
    recipe: data.fetchRecipesResponse.recipes[0],
    upvote: jest.fn(),
    downvote: jest.fn(),
  };

  const wrapper = shallow(<RecipeItem {...props} />);

  return {
    wrapper,
    props
  };
};

const { wrapper, props } = setup();

describe('RecipeItem', () => {
  it('should render correctly',
    (done) => {
      expect(toJson(wrapper)).toMatchSnapshot();
      done();
    });

  it('should display the recipe view counts', (done) => {
    expect(wrapper.find('.card-title').first().text()).toEqual('0 Views');
    done();
  });

  it('should show the recipe name', (done) => {
    expect(wrapper.find('.card-title').last().text()).toEqual('Chicken Soup');
    done();
  });

  describe('onClick', () => {
    it('of the upvote button should upvote the recipe', (done) => {
      const upVoteButton = wrapper.find('.like');
      upVoteButton.simulate('click');
      expect(props.upvote.mock.calls.length).toEqual(1);
      done();
    });

    it('of the downvote button should downvote the recipe', (done) => {
      const downVoteButton = wrapper.find('.unlike');
      downVoteButton.simulate('click');
      expect(props.downvote.mock.calls.length).toEqual(1);
      done();
    });
  });
});
