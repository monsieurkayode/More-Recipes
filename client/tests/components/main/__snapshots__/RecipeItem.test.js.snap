// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeItem should render correctly 1`] = `
<div
  className="col m4"
>
  <div
    className="card hoverable views-small"
  >
    <div
      className="card-image"
    >
      <Link
        replace={false}
        to="/recipes/1"
      >
        <img
          alt="Chicken Soup"
          src="spice.jpg"
        />
      </Link>
      <span
        className="card-title"
      >
        0
         Views
      </span>
    </div>
    <div
      className="card-content small-cards-home"
    >
      <div
        className="card-title truncate"
      >
        <strong>
          Chicken Soup
        </strong>
      </div>
      <div
        className="card-details"
      >
        <Link
          className="chip teal white-text"
          replace={false}
          to="/category/specials"
        >
          Specials
        </Link>
        <span>
          <div
            className="unlike right"
            onClick={[Function]}
          >
            <i
              className="material-icons tiny reaction"
            >
              thumb_down
            </i>
            0
          </div>
          <div
            className="like right"
            onClick={[Function]}
          >
            <i
              className="material-icons tiny reaction"
            >
              thumb_up
            </i>
            0
          </div>
        </span>
      </div>
    </div>
  </div>
</div>
`;
